### PR TITLE
Rule to check for direct usages of GORM methods

### PR DESCRIPTION
**Pull request by musketyr**
It is very difficult to upgrade to newer versions of Grails when you have the persistence logic spread across the whole application. This rule checks for direct usages of GORM instance and static methods so it's easier to find them.

### CODE REVIEW

File: docs/codenarc-rules-grails.md [hunk 1]
```diff
@@ -5,6 +5,59 @@ title: CodeNarc - Grails Rules
 
 # Grails Rules  ("*rulesets/grails.xml*")
 
+## GrailsDomainGormMethods Rule
+
+<Since CodeNarc 1.6>
+
+Database operation should be performed by Data Services instead of calling GORM static and instance methods.
+
```
Comment (chrismair) [inline docs/codenarc-rules-grails.md:13]: Is Grails Data Services available in all version of Grails? If not, please indicate for which versions of Grails this rule applies. Also, how about adding a link to http://gorm.grails.org/latest/hibernate/manual/index.html#dataServices.

File: gradlew.bat [hunk 2]
```diff
@@ -1,100 +1,100 @@
-@rem
```
Comment (chrismair) [inline gradlew.bat:1]: Please back out your changes to gradlew.bat.

File: src/main/groovy/org/codenarc/rule/grails/GrailsDomainGormMethodsRule.groovy [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            'withTenant',
+            'withTransaction',
+    ].asImmutable()
+
+    String name = 'GrailsDomainGormMethods'
+    int priority = 3
+    Class astVisitorClass = GrailsDomainGormMethodsAstVisitor
+    int compilerPhase = Phases.SEMANTIC_ANALYSIS
+    List<String> gormStaticMethodsNamesList = DEFAULT_GORM_STATIC_METHOD_NAMES
+
```
Comment (chrismair) [inline src/main/groovy/org/codenarc/rule/grails/GrailsDomainGormMethodsRule.groovy:92]: Is the `List<String> getGormStaticMethodsNamesList()` method necessary? It is a property already.

File: src/test/groovy/org/codenarc/rule/grails/GrailsDomainGormMethodsRuleTest.groovy [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+class GrailsDomainGormMethodsRuleTest extends AbstractRuleTestCase<GrailsDomainGormMethodsRule> {
+
+    @Test
+    void test_RuleProperties() {
+        assert rule.priority == 3
+        assert rule.name == 'GrailsDomainGormMethods'
+    }
+
+    @Test
+    void test_SomeCondition_NoViolations() {
```
Comment (chrismair) [inline src/test/groovy/org/codenarc/rule/grails/GrailsDomainGormMethodsRuleTest.groovy:35]: Please rename all test methods to remove "SomeCondition" or replace it with something more specific.

File: src/test/groovy/org/codenarc/rule/grails/GrailsDomainGormMethodsRuleTest.groovy [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                    messageText: 'Prefer GORM Data Services to GORM static calls like \'list\'',
+            ],
+            [
+                    lineNumber: 26,
+                    sourceLineText: 'firstEntity.save()',
+                    messageText: 'Prefer GORM Data Services to GORM instance calls like \'save\'',
+            ]
+        )
+    }
+
```
Comment (chrismair) [inline src/test/groovy/org/codenarc/rule/grails/GrailsDomainGormMethodsRuleTest.groovy:111]: Please add two tests: setting/validating `setGormStaticMethodsNames`; and setting `gormStaticMethodsNamesList` directly.

File: docs/codenarc-rules-grails.md [hunk 6]
```diff
@@ -5,6 +5,59 @@ title: CodeNarc - Grails Rules
...
+Using the GORM static and instance methods may lead to spreading the persistence logic across the whole
+application instead of concentrating it into services. It makes difficult to find all the code working
+with the database in case of upgrades to the newer versions of Grails which require all persistence code
+running inside transactions.  
+
```
Comment (chrismair) [inline docs/codenarc-rules-grails.md:18]: Add the standard text and link to Enhanced Rules page.

See CloneWithoutCloneable Rule description within "codenarc-rules-enhanced.md" -- copy the paragraph that starts with "NOTE: This is a [CodeNarc Enhanced Classpath Rule]...".

File: src/main/resources/rulesets/enhanced.xml [hunk 7]
```diff
@@ -11,6 +11,7 @@
         any referenced classes, on the classpath.
```
Comment (chrismair) [inline src/main/resources/rulesets/enhanced.xml:11]: Please move your new rule (back) into the grails ruleset. You document it within that ruleset markdown file, and the rule makes more sense there. We do have other *enhanced* rules in other rulesets, so it is okay to not be in the enhanced ruleset.

File: gradlew.bat (hunk 2 above)
Comment (musketyr) [reply gradlew.bat:1]: I'm bit lost here. Either IntelliJ or Git is reformatting it so I can't push the changes.

**Comment by musketyr** [comment]
@chrismair all the comments should be fixed. I had to open a new PR because I wasn't able to restore the Gradle Wrapper on this branch.

please, let's meet here #515 for the follow up

**Pull request closed without merging**